src/psyclone/nemo.py

```diff
@@ -0,0 +1,954 @@
+                :py:class:`fparser.two.Fortran2003.Block_Nonlabel_Do_Construct`
+
+        :raises InternalError: if the supplied loop node is not recognised.
+        '''
+        from fparser.two.Fortran2003 import Block_Nonlabel_Do_Construct, \
+            Assignment_Stmt
+
+        if isinstance(loop, Block_Nonlabel_Do_Construct):
+            self._load_from_loop(loop)
+        elif isinstance(loop, Assignment_Stmt):
+            self._load_from_implicit_loop(loop)
+        else:
+            raise InternalError(
+                "Expecting either Block_Nonlabel_Do_Construct or "
+                "Assignment_Stmt but got {0}".format(str(type(loop))))
+
+    def _load_from_loop(self, loop):
+        '''
+        Populate the state of this NemoKern object from an fparser2
+        AST for an explicit loop.
+
+        :param loop: Node in the fparser2 AST representing an implicit loop.
+        :type loop: :py:class:`fparser.two.Fortran2003.Assignment_Stmt`
+
+        :raises InternalError: if first child of supplied loop node is not a \
+                           :py:class:`fparser.two.Fortran2003.Nonlabel_Do_Stmt`
+        '''
+        from fparser.two.Fortran2003 import Nonlabel_Do_Stmt, End_Do_Stmt
+
+        # Keep a pointer to the original loop in the AST
+        self._loop = loop
+
+        if not isinstance(loop.content[0], Nonlabel_Do_Stmt):
+            raise InternalError("Expecting Nonlabel_Do_Stmt as first child "
+                                "of Block_Nonlabel_Do_Construct but "
+                                "got {0}".format(type(loop.content[0])))
+        self._body = []
+        for content in loop.content[1:]:
+            if isinstance(content, End_Do_Stmt):
+                break
+            self._body.append(content)
+
+        #  I could get this by walking back up the tree and counting how
+        #  many NemoLoops I have as ancestors before I come across
+        #  something that is not a loop...
+        # if len(self._loop_vars) == 2:
+        #     self._kernel_type = "2D"
+        # else:
+        #     self._kernel_type = "3D"
+
+        # TODO bring habakkuk up-to-date with changes to fparser and then
+        # uncomment this code.
+
+        #  Analyse the loop body to identify private and shared variables
+        # from habakkuk.make_dag import dag_of_code_block
+        #  Create a DAG of the kernel code block using Habakkuk
+        # kernel_dag = dag_of_code_block(loop, "nemo_kernel")
+        # inputs = kernel_dag.input_nodes()
+        # outputs = kernel_dag.output_nodes()
+        # print "Kernel has {0} outputs: ".format(len(outputs)) + \
+        #     ",".join([node.variable.orig_name for node in outputs])
+        self._shared_vars = set()
+        self._first_private_vars = set()
+        self._private_vars = set()
+        #  If there are scalar variables that are inputs to the DAG (other than
+        #  the loop counters) then they must be declared first-private.
```

Comment:
Please add "in an OpenMP loop directive."

Reply:
Done.